### PR TITLE
Add performance tests for creating and validating a MAX_BLOCK_SIZE transaction

### LINKED ISSUE #971: Create worst-case-block-verification time tests.
Before we land #962 (which closes #765), let's add two performance tests:

Test A - measure latency/memory-usage for verifying a very large transparent (Bitcoin) transaction's signatures. I think this is a transaction with as many valid SIGHASH_ALL txin signatures as can fit in a block. Note that the test must ensure the txn is valid in the block, so for example introducing a new limit like ``MAX_BLOCK_SIGHASH`` [1] may break this test, which would need to be adjusted to fit the new limit.

~~Test B - measure latency/memory-usage for verifying a block full of JoinSplits. I don't think it matters if these are in a single txn or separate txns, although the non-JS overhead might be lower if they're in a single txn.~~

With these in place, we can track performance changes as we change limit constants or introduce new limits. (Side note: we view introducing new limits with disdain.)

My expectations are that Test A has super-linear time w/ sig count and Test B has linear size w/ sig count.

Notice: other details aside from limits may affect these measurements. For example, if we change JoinSplit verification from serial to parallel, the latency may drop.

[1] http://gavinandresen.ninja/a-guided-tour-of-the-2mb-fork

**Edit:** Let's remove the JoinSplit test from this ticket. JoinSplit verification is constant time, and highly parallelizeable, so we expect worse case verification time to scale linearly with block size.

### PR DESCRIPTION
**Pull request by str4d**
Closes #971.

### CODE REVIEW

**Comment by ebfull** [comment]
Incredible work on this, reviewing now.

File: src/zcbenchmarks.cpp [hunk 1]
```diff
@@ -126,3 +132,121 @@ double benchmark_verify_equihash()
...
 }
 
+double benchmark_large_tx(bool testValidate)
+{
+    // Note that the transaction size is checked against the splitting
+    // transaction, not the merging one. Thus we are assuming that transactions
+    // with 1 input and N outputs are about the same size as transactions with
+    // N inputs and 1 output.
+    mapArgs["-blockmaxsize"] = itostr(MAX_BLOCK_SIZE);
+    int nBlockSizeRemaining = MAX_BLOCK_SIZE-1000;
```
Comment (ebfull) [inline src/zcbenchmarks.cpp]: This has a confusing name, since it never decreases.

**Comment by ebfull** [comment]
This should also clear the mempool before it runs, just in case another test interferes.

**Comment by ebfull** [comment]
ACK, no blocking nits.

File: src/zcbenchmarks.cpp (hunk 1 above)
Comment (str4d) [inline src/zcbenchmarks.cpp]: Leftover from the previous approach, will fix

File: src/zcbenchmarks.cpp [hunk 2]
```diff
@@ -126,3 +132,122 @@ double benchmark_verify_equihash()
...
+        mtx2.vin[i].prevout.hash = hash;
+        mtx2.vin[i].prevout.n = i;
+    }
+
+    for (int i = 0; i < mtx2.vin.size(); i++) {
+        SignSignature(*pwalletMain, mtx, mtx2, i);
+    }
+
+    double ret = timer_stop();
+    if (!testValidate)
```
Comment (defuse) [inline src/zcbenchmarks.cpp]: `pblocktemplate` doesn't get freed in this branch.

**Comment by defuse** [comment]
NACK because of the memory leak thing but after that's fixed it looks good.

**Comment by str4d** [comment]
Just before merging this, we need to update bbotzc to run these tests (so that this PR will have the corresponding data).

File: src/zcbenchmarks.cpp [hunk 3]
```diff
@@ -126,3 +132,122 @@ double benchmark_verify_equihash()
...
+        EhOptimisedSolve(n, k, curr_state, solns);
+        for (auto soln : solns) {
+            pblock->nSolution = soln;
+            if (UintToArith256(pblock->GetHash()) > hashTarget) {
+                continue;
+            }
+            goto processblock;
+        }
+        pblock->nNonce = ArithToUint256(UintToArith256(pblock->nNonce) + 1);
+    }
```
Comment (defuse) [inline src/zcbenchmarks.cpp]: Any way we could reduce the duplication here? This is basically the miner code right?

Comment (str4d) [inline src/zcbenchmarks.cpp]: Not at this stage, but maybe in future once we refactor `miner.cpp`.

**Pull request merged**